routers/init.go

```diff
@@ -45,23 +49,49 @@ import (
 	"gitea.com/go-chi/session"
 )
 
-// NewServices init new services
-func NewServices() {
-	setting.NewServices()
-	if err := storage.Init(); err != nil {
-		log.Fatal("storage init failed: %v", err)
+func mustInit(fn func() error) {
+	err := fn()
+	if err != nil {
+		ptr := reflect.ValueOf(fn).Pointer()
+		fi := runtime.FuncForPC(ptr)
+		log.Fatal("%s failed: %v", fi.Name(), err)
 	}
-	if err := repository.NewContext(); err != nil {
-		log.Fatal("repository init failed: %v", err)
+}
+
+func mustInitCtx(ctx context.Context, fn func(ctx context.Context) error) {
+	err := fn(ctx)
+	if err != nil {
+		ptr := reflect.ValueOf(fn).Pointer()
+		fi := runtime.FuncForPC(ptr)
+		log.Fatal("%s(ctx) failed: %v", fi.Name(), err)
 	}
-	mailer.NewContext()
-	if err := cache.NewContext(); err != nil {
-		log.Fatal("Unable to start cache service: %v", err)
+}
+
+// InitGitServices init new services for git, this is also called in `contrib/pr/checkout.go`
+func InitGitServices() {
+	setting.NewServices()
+	mustInit(storage.Init)
+	mustInit(repository.NewContext)
+}
+
+func syncAppPathForGit(ctx context.Context) (err error) {
+	runtimeState := new(appstate.RuntimeState)
+	if err = setting.AppState.Get(runtimeState); err != nil {
```

Comment:
Imho this looks much cleaner:

```suggestion
	runtimeState, err := setting.AppState.Get()
	if err != nil {
```

Reply:
As above